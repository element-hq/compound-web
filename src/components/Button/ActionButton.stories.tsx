/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { type Meta, type StoryObj } from "@storybook/react-vite";
import { fn } from "storybook/test";

import * as icons from "@vector-im/compound-design-tokens/assets/web/icons";

import { Button as ButtonComponent } from "./Button";
import { Tooltip } from "../Tooltip/Tooltip";
import { TooltipProvider } from "../Tooltip/TooltipProvider";

const Template: React.FC<
  { label: string; as: "button" | "a" } & React.ComponentProps<
    typeof ButtonComponent
  >
> = ({ label, ...args }) => (
  <TooltipProvider>
    <Tooltip label={label}>
      <ButtonComponent iconOnly {...args} />
    </Tooltip>
  </TooltipProvider>
);

const meta = {
  title: "Button/Actions",
  component: Template,
  tags: ["autodocs"],
  argTypes: {
    label: {
      type: "string",
    },
    size: {
      options: ["sm", "lg"],
      control: { type: "inline-radio" },
    },
    kind: {
      options: ["primary", "secondary", "tertiary"],
      control: { type: "inline-radio" },
    },
    destructive: { type: "boolean" },
    disabled: {
      type: "boolean",
    },
    as: {
      options: ["a", "button"],
      control: { type: "inline-radio" },
    },
    Icon: {
      options: Object.keys(icons),
      mapping: icons,
      control: { type: "select" },
    },
    onClick: { action: "onClick" },
  },
  args: {
    label: "Action",
    size: "lg",
    as: "button",
    destructive: false,
    disabled: false,
    onClick: fn(),
  },
  parameters: {
    controls: {
      exclude: ["children"],
    },
  },
} satisfies Meta<typeof Template>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Settings: Story = {
  args: {
    label: "Settings",
    kind: "secondary",
    Icon: icons.SettingsSolidIcon,
  },
};

export const ShareLink: Story = {
  args: {
    label: "Copy link",
    kind: "secondary",
    Icon: icons.LinkIcon,
  },
};

export const VideoOn: Story = {
  args: {
    label: "Turn video off",
    kind: "secondary",
    Icon: icons.VideoCallSolidIcon,
  },
};

export const VideoOff: Story = {
  args: {
    label: "Turn video on",
    kind: "primary",
    Icon: icons.VideoCallOffSolidIcon,
  },
};

export const MicOn: Story = {
  args: {
    label: "Turn mic off",
    kind: "secondary",
    Icon: icons.MicOnSolidIcon,
  },
};

export const MicOff: Story = {
  args: {
    label: "Turn mic on",
    kind: "primary",
    Icon: icons.MicOffSolidIcon,
  },
};

export const ShareScreen: Story = {
  args: {
    label: "Share screen",
    kind: "secondary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const SharingScreen: Story = {
  args: {
    label: "Stop sharing screen",
    kind: "primary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const Fullscreen: Story = {
  args: {
    label: "Full screen",
    kind: "secondary",
    Icon: icons.ExpandIcon,
  },
};

export const EndCall: Story = {
  args: {
    label: "End call",
    kind: "primary",
    destructive: true,
    Icon: icons.EndCallIcon,
  },
};

export const Invite: Story = {
  args: {
    label: "Invite",
    kind: "primary",
    size: "sm",
    Icon: icons.UserAddIcon,
  },
};

export const Edit: Story = {
  args: {
    label: "Edit",
    kind: "primary",
    size: "sm",
    Icon: icons.EditIcon,
  },
};

export const Search: Story = {
  args: {
    label: "Search",
    kind: "secondary",
    size: "sm",
    Icon: icons.SearchIcon,
  },
};

export const Save: Story = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
  },
};

export const SaveDisabled: Story = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
    disabled: true,
  },
};

export const Cancel: Story = {
  args: {
    label: "Cancel",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};

export const Close: Story = {
  args: {
    label: "Close",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};
