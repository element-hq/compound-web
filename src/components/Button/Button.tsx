/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from "classnames";
import React, {
  ComponentType,
  PropsWithChildren,
  forwardRef,
  ForwardedRef,
  Ref,
} from "react";
import styles from "./Button.module.css";
import { UnstyledButton, UnstyledButtonPropsFor } from "./UnstyledButton";

interface ButtonComponent {
  // With the explicit `as` prop
  <C extends React.ElementType>(
    props: { as: C } & ButtonPropsFor<C>,
  ): React.ReactElement;
  // Without the explicit `as` prop, defaulting to a <button>
  (props: ButtonPropsFor<"button">): React.ReactElement;
}

type ButtonOwnProps = PropsWithChildren<{
  /**
   * The type of button.
   */
  kind?: "primary" | "secondary" | "tertiary" | "destructive"; // TODO: Refine the naming
  /**
   * The t-shirt size of the button.
   */
  size?: "sm" | "lg";
  /**
   * An icon to display within the button.
   */
  Icon?: ComponentType<React.SVGAttributes<SVGElement>>;
}>;

type ButtonPropsFor<C extends React.ElementType> = ButtonOwnProps &
  UnstyledButtonPropsFor<C>;

/**
 * A button component that can be transformed into a link, but keep the button
 * styling using the `as` property.
 */
export const Button = forwardRef(function Button<
  C extends React.ElementType = "button",
>(
  {
    as,
    kind = "primary",
    size = "lg",
    children,
    className,
    Icon,
    disabled,
    ...props
  }: ButtonPropsFor<C> & { as?: C },
  ref: ForwardedRef<C>,
): React.ReactElement {
  const classes = classNames(styles.button, className, {
    [styles["has-icon"]]: Icon,
  });

  return (
    <UnstyledButton
      {...props}
      as={as || ("button" as const)}
      ref={ref as Ref<C>}
      className={classes}
      data-size={size}
      data-kind={kind}
      tabIndex={0}
      disabled={disabled}
    >
      {Icon && (
        <Icon
          width={20}
          height={20}
          className={styles.icon}
          aria-hidden={true}
        />
      )}
      {children}
    </UnstyledButton>
  );
}) as ButtonComponent;
