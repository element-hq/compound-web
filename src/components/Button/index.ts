/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

export { Button } from "./Button";
export { IconButton } from "./IconButton/IconButton";
