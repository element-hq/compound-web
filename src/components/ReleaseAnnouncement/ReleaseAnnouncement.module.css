/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

.content {
  /* 320px + 20px padding left + 20px padding right */
  max-inline-size: 320px;
  border-radius: var(--cpd-space-3x);
  background: var(--cpd-color-bg-action-primary-rest);
  padding: var(--cpd-space-5x) var(--cpd-space-6x);

  /* from figma */
  box-shadow: 0 4px 24px 0 rgb(0 0 0 / 10%);

  /*
  * Grid, we want to have the following layout:
  * ----------------------------------
  * -  ---------------   ----------  -
  * -  | heading     |   |        |  -
  * -  --------------    | button |  -
  * -  ---------------   | button |  -
  * -  | description |   |        |  -
  * -  ---------------   ----------  -
  * ----------------------------------
  */
  display: grid;
  grid-template:
    "header button" auto
    "description button" auto;
  align-items: center;
  column-gap: var(--cpd-space-6x);
}

.header {
  margin: 0;
  color: var(--cpd-color-text-on-solid-primary);
  grid-area: header;
}

.description {
  color: var(--cpd-color-gray-500);
  grid-area: description;
}

.button {
  color: var(--cpd-color-bg-subtle-secondary);
  grid-area: button;
}

.arrow {
  fill: var(--cpd-color-bg-action-primary-rest);
}
