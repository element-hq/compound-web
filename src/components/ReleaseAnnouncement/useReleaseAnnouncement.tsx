/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import {
  arrow,
  autoUpdate,
  limitShift,
  offset,
  Placement,
  shift,
  useFloating,
  useId,
  useInteractions,
  useRole,
} from "@floating-ui/react";
import { MouseEventHandler, useMemo, useRef } from "react";

interface UseReleaseAnnouncementProps {
  /**
   * Whether the release announcement is open.
   */
  open: boolean;
  /**
   * The header of the release announcement.
   */
  header: string;
  /**
   * The description of the release announcement.
   */
  description: string;
  /**
   * The label of the close button.
   */
  closeLabel: string;
  /**
   * The placement of the release announcement.
   */
  placement: Placement;
  /**
   * The event handler for the close button.
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
  /**
   * Whether to display an arrow.
   */
  displayArrow?: boolean;
}

/**
 * This hook provides the necessary props to create a release announcement.
 */
export function useReleaseAnnouncement({
  open,
  header,
  description,
  closeLabel,
  placement,
  onClick,
  displayArrow,
}: UseReleaseAnnouncementProps) {
  // Set on `aria-labelledby` attribute
  const labelId = useId();
  // Set on `aria-describedby` attribute
  const descriptionId = useId();
  const arrowRef = useRef(null);

  const data = useFloating({
    placement,
    open,
    whileElementsMounted: autoUpdate,
    middleware: [
      // arrow height 12px + 4px padding
      offset(16),
      shift({ limiter: limitShift({ offset: 50 }) }),
      // add the little arrow along with the floating content
      displayArrow &&
        arrow({
          element: arrowRef,
        }),
    ],
  });

  // The role is used to determine the correct aria role for the floating content.
  const role = useRole(data.context);
  const interactions = useInteractions([role]);

  // Put in the context all the useful data and interactions
  return useMemo(
    () => ({
      open,
      ...data,
      ...interactions,
      labelId,
      descriptionId,
      header,
      description,
      closeLabel,
      onClick,
      displayArrow,
      arrowRef,
    }),
    [
      open,
      role,
      interactions,
      data,
      labelId,
      descriptionId,
      header,
      description,
      closeLabel,
      onClick,
      displayArrow,
      arrowRef,
    ],
  );
}
