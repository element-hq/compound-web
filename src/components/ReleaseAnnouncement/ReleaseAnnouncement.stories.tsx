/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";
import { Button } from "../Button";
import { ReleaseAnnouncement as ReleaseAnnouncementComponent } from "./ReleaseAnnouncement";

type Props = Omit<
  React.ComponentProps<typeof ReleaseAnnouncementComponent>,
  "open" | "children" | "onClick"
>;

const Template: React.FC<Props> = (args) => {
  const [open, setOpen] = useState(true);

  return (
    <ReleaseAnnouncementComponent
      {...args}
      open={open}
      onClick={() => setOpen(false)}
    >
      <Button onClick={() => setOpen((open) => !open)}>Open menu</Button>
    </ReleaseAnnouncementComponent>
  );
};

const meta = {
  title: "Release Announcement",
  component: Template,
  tags: ["autodocs"],
  argTypes: {},
  args: {
    header: "Notifications have moved",
    description: "From now on, click the icon here to view your notifications.",
    placement: "right",
    closeLabel: "Ok",
  },
} satisfies Meta<typeof Template>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = { args: {} };

export const MultiLinesContent: Story = {
  args: {
    header: "A multi line header which is taking multiple lines",
    description:
      "A multi line description which is taking multiple lines. The OK button should be vertically centred.",
  },
};

export const BottomPlacement: Story = {
  args: {
    placement: "bottom",
  },
};

export const NoArrow: Story = {
  args: {
    displayArrow: false,
  },
};
