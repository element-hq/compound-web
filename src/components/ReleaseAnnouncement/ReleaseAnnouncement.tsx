/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, {
  cloneElement,
  isValidElement,
  type JSX,
  type PropsWithChildren,
  type Ref,
} from "react";
import {
  FloatingPortal,
  FloatingFocusManager,
  type Placement,
  useMergeRefs,
  FloatingArrow,
} from "@floating-ui/react";
import { Text } from "../Typography/Text";
import { Button } from "../Button";
import styles from "./ReleaseAnnouncement.module.css";
import {
  useReleaseAnnouncementContext,
  ReleaseAnnouncementContext,
} from "./ReleaseAnnouncementContext";
import { useReleaseAnnouncement } from "./useReleaseAnnouncement";

type UseReleaseAnnouncementParam = Parameters<typeof useReleaseAnnouncement>[0];

interface ReleaseAnnouncementProps
  extends Omit<UseReleaseAnnouncementParam, "placement" | "displayArrow"> {
  /**
   * The placement of the component
   * @default "right"
   */
  placement?: Placement;
  /**
   * Whether to display an arrow.
   * @default true
   */
  displayArrow?: boolean;
}

/**
 * The ReleaseAnnouncement component purpose is to inform the user of a new available feature.
 * This component is a floating component that will appear next to an anchor.
 * @param children - Act as an anchor, the component will be displayed alongside of it.
 * @param placement - The placement of the component
 */
export function ReleaseAnnouncement({
  /**
   * The children anchor should be a single valid React element.
   */
  children,
  placement = "right",
  displayArrow = true,
  ...props
}: PropsWithChildren<ReleaseAnnouncementProps>): JSX.Element {
  const context = useReleaseAnnouncement({ placement, displayArrow, ...props });

  return (
    <ReleaseAnnouncementContext.Provider value={context}>
      <ReleaseAnnouncementAnchor>{children}</ReleaseAnnouncementAnchor>
      <ReleaseAnnouncementContainer>
        <ReleaseAnnouncementContent />
      </ReleaseAnnouncementContainer>
    </ReleaseAnnouncementContext.Provider>
  );
}

/**
 * The anchor for the ReleaseAnnouncement components.
 * The Release Announcement will appear next to this element.
 * @param children - should be a single valid React element
 * @constructor
 */
function ReleaseAnnouncementAnchor({
  children,
}: Readonly<PropsWithChildren>): JSX.Element {
  const context = useReleaseAnnouncementContext();

  // The children can have a ref and we don't want to discard it
  // Doing a dirty cast to get the optional ref
  const childrenRef = (children as unknown as { ref?: Ref<HTMLElement> })?.ref;
  const ref = useMergeRefs([context.refs.setReference, childrenRef]);

  if (!isValidElement(children)) {
    throw new Error(
      "ReleaseAnnouncement anchor must be a single valid React element",
    );
  }

  const referenceProps = context.getReferenceProps({
    ref,
    // If the ReleaseAnnouncement is open, we need manually aria-describedby.
    // The RA has the dialog role and it's not adding automatically the aria-describedby.
    ...(context.open && {
      "aria-describedby": context.getFloatingProps().id as string,
    }),
  });

  // getReferenceProps includes aria-expanded (being true when the popup is shown)
  // but axe only allows this on certain elements and we use release anncounements on
  // all sorts of elements. Semantically, I don't think it makes sense because the point
  // of aria-expanded is a hint to the user that there is something that can be expanded.
  // The user can't cause a release announcement to be shown, it just appears when it thinks
  // it's time, so I think adding it here is misleading.
  delete referenceProps["aria-expanded"];

  return cloneElement(children, referenceProps);
}

/**
 * The container for the ReleaseAnnouncement components.
 * Manage focus and positioning of the children.
 * @param children
 */
function ReleaseAnnouncementContainer({
  children,
}: PropsWithChildren): JSX.Element | null {
  const {
    context: floatingContext,
    arrowRef,
    displayArrow,
    ...rest
  } = useReleaseAnnouncementContext();

  if (!floatingContext.open) return null;

  return (
    <FloatingPortal>
      <FloatingFocusManager context={floatingContext} modal={false}>
        <div
          ref={rest.refs.setFloating}
          style={rest.floatingStyles}
          aria-labelledby={rest.labelId}
          aria-describedby={rest.descriptionId}
          {...rest.getFloatingProps()}
          className={styles.content}
        >
          {displayArrow && (
            <FloatingArrow
              ref={arrowRef}
              context={floatingContext}
              // design absolute value
              width={20}
              height={12}
              className={styles.arrow}
            />
          )}
          {children}
        </div>
      </FloatingFocusManager>
    </FloatingPortal>
  );
}

/**
 * The content of the ReleaseAnnouncement component.
 * ---------------------------------------------------------------
 * - The header in a bold text                  -------------    -
 * - A description in a regular text           | Close button |  -
 * - Description can be on multiple lines       -------------    -
 * ---------------------------------------------------------------
 */
function ReleaseAnnouncementContent(): JSX.Element {
  const { labelId, descriptionId, header, description, closeLabel, onClick } =
    useReleaseAnnouncementContext();

  return (
    <>
      <Text
        as="h3"
        id={labelId}
        className={styles.header}
        size="lg"
        weight="semibold"
      >
        {header}
      </Text>
      <Text
        as="span"
        id={descriptionId}
        className={styles.description}
        size="sm"
        weight="regular"
      >
        {description}
      </Text>
      <Button
        size="sm"
        kind="secondary"
        className={styles.button}
        onClick={onClick}
      >
        {closeLabel}
      </Button>
    </>
  );
}
