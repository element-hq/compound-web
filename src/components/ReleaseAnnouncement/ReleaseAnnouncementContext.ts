/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { createContext, useContext } from "react";
import { useReleaseAnnouncement } from "./useReleaseAnnouncement";

type ContextType = ReturnType<typeof useReleaseAnnouncement> | null;
/**
 * The context for the ReleaseAnnouncement components.
 */
export const ReleaseAnnouncementContext = createContext<ContextType>(null);

/**
 * Provides the context for the ReleaseAnnouncement components.
 */
export function useReleaseAnnouncementContext() {
  const context = useContext(ReleaseAnnouncementContext);

  if (context == null) {
    throw new Error(
      "ReleaseAnnouncement components must be wrapped in <ReleaseAnnouncement />",
    );
  }

  return context;
}
