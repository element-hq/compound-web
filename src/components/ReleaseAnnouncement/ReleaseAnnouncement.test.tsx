/*
 * Copyright 2024 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import { composeStories } from "@storybook/react";
import * as stories from "./ReleaseAnnouncement.stories";
import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import { ReleaseAnnouncement } from "./ReleaseAnnouncement";
import { Button } from "../Button";

const { Default, MultiLinesContent, BottomPlacement, NoArrow } =
  composeStories(stories);

describe("ReleaseAnnouncement", () => {
  it("renders", async () => {
    render(<Default />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders with a long label header and description", async () => {
    render(<MultiLinesContent />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders the component at the bottom of the trigger button", async () => {
    render(<BottomPlacement />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders the component when closed", async () => {
    render(
      <ReleaseAnnouncement
        open={false}
        header="header"
        description="description"
        closeLabel="Ok"
        onClick={vi.fn()}
      >
        <Button>Open</Button>
      </ReleaseAnnouncement>,
    );
    expect(screen.queryByRole("dialog")).toBeNull();
  });

  it("renders without an arrow", async () => {
    render(<NoArrow />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });
});
