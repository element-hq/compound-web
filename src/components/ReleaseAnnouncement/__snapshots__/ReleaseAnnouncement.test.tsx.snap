// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReleaseAnnouncement > renders 1`] = `
<div
  aria-describedby="«r1»"
  aria-labelledby="«r0»"
  class="_content_4b2dd5"
  data-floating-ui-focusable=""
  id="«r2»"
  role="dialog"
  style="position: absolute; left: 0px; top: 0px; transform: translate(16px, 50px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_4b2dd5"
    height="20"
    style="position: absolute; pointer-events: none; right: calc(100% - 0px); transform: rotate(90deg); top: -1px;"
    viewBox="0 0 20 20"
    width="20"
  >
    <path
      d="M0,0 H20 L10,12 Q10,12 10,12 Z"
      stroke="none"
    />
    <clippath
      id="«r5»"
    >
      <rect
        height="20"
        width="20"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <h3
    class="_typography_61e060 _font-body-lg-semibold_61e060 _header_4b2dd5"
    id="«r0»"
  >
    Notifications have moved
  </h3>
  <span
    class="_typography_61e060 _font-body-sm-regular_61e060 _description_4b2dd5"
    id="«r1»"
  >
    From now on, click the icon here to view your notifications.
  </span>
  <button
    class="_button_0829c1 _button_4b2dd5"
    data-kind="secondary"
    data-size="sm"
    role="button"
    tabindex="0"
  >
    Ok
  </button>
</div>
`;

exports[`ReleaseAnnouncement > renders the component at the bottom of the trigger button 1`] = `
<div
  aria-describedby="«rd»"
  aria-labelledby="«rc»"
  class="_content_4b2dd5"
  data-floating-ui-focusable=""
  id="«re»"
  role="dialog"
  style="position: absolute; left: 0px; top: 0px; transform: translate(50px, 16px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_4b2dd5"
    height="20"
    style="position: absolute; pointer-events: none; bottom: 100%; transform: rotate(180deg); left: -1px;"
    viewBox="0 0 20 20"
    width="20"
  >
    <path
      d="M0,0 H20 L10,12 Q10,12 10,12 Z"
      stroke="none"
    />
    <clippath
      id="«rh»"
    >
      <rect
        height="20"
        width="20"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <h3
    class="_typography_61e060 _font-body-lg-semibold_61e060 _header_4b2dd5"
    id="«rc»"
  >
    Notifications have moved
  </h3>
  <span
    class="_typography_61e060 _font-body-sm-regular_61e060 _description_4b2dd5"
    id="«rd»"
  >
    From now on, click the icon here to view your notifications.
  </span>
  <button
    class="_button_0829c1 _button_4b2dd5"
    data-kind="secondary"
    data-size="sm"
    role="button"
    tabindex="0"
  >
    Ok
  </button>
</div>
`;

exports[`ReleaseAnnouncement > renders with a long label header and description 1`] = `
<div
  aria-describedby="«r7»"
  aria-labelledby="«r6»"
  class="_content_4b2dd5"
  data-floating-ui-focusable=""
  id="«r8»"
  role="dialog"
  style="position: absolute; left: 0px; top: 0px; transform: translate(16px, 50px);"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="_arrow_4b2dd5"
    height="20"
    style="position: absolute; pointer-events: none; right: calc(100% - 0px); transform: rotate(90deg); top: -1px;"
    viewBox="0 0 20 20"
    width="20"
  >
    <path
      d="M0,0 H20 L10,12 Q10,12 10,12 Z"
      stroke="none"
    />
    <clippath
      id="«rb»"
    >
      <rect
        height="20"
        width="20"
        x="0"
        y="0"
      />
    </clippath>
  </svg>
  <h3
    class="_typography_61e060 _font-body-lg-semibold_61e060 _header_4b2dd5"
    id="«r6»"
  >
    A multi line header which is taking multiple lines
  </h3>
  <span
    class="_typography_61e060 _font-body-sm-regular_61e060 _description_4b2dd5"
    id="«r7»"
  >
    A multi line description which is taking multiple lines. The OK button should be vertically centred.
  </span>
  <button
    class="_button_0829c1 _button_4b2dd5"
    data-kind="secondary"
    data-size="sm"
    role="button"
    tabindex="0"
  >
    Ok
  </button>
</div>
`;

exports[`ReleaseAnnouncement > renders without an arrow 1`] = `
<div
  aria-describedby="«rn»"
  aria-labelledby="«rm»"
  class="_content_4b2dd5"
  data-floating-ui-focusable=""
  id="«ro»"
  role="dialog"
  style="position: absolute; left: 0px; top: 0px; transform: translate(16px, 50px);"
  tabindex="-1"
>
  <h3
    class="_typography_61e060 _font-body-lg-semibold_61e060 _header_4b2dd5"
    id="«rm»"
  >
    Notifications have moved
  </h3>
  <span
    class="_typography_61e060 _font-body-sm-regular_61e060 _description_4b2dd5"
    id="«rn»"
  >
    From now on, click the icon here to view your notifications.
  </span>
  <button
    class="_button_0829c1 _button_4b2dd5"
    data-kind="secondary"
    data-size="sm"
    role="button"
    tabindex="0"
  >
    Ok
  </button>
</div>
`;
